# State Objective

The objective of my analysis is to:
1. Identify the variables that most strongly impacted the likelihood of survival
2. To create a model that predicts survival based on these variables

In [1]:
import pandas as pd 

Location = "titanic3.xls"
df= pd.read_excel(Location)

# Clean Data

In [2]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1309 entries, 0 to 1308
Data columns (total 14 columns):
pclass       1309 non-null int64
survived     1309 non-null int64
name         1309 non-null object
sex          1309 non-null object
age          1046 non-null float64
sibsp        1309 non-null int64
parch        1309 non-null int64
ticket       1309 non-null object
fare         1308 non-null float64
cabin        295 non-null object
embarked     1307 non-null object
boat         486 non-null object
body         121 non-null float64
home.dest    745 non-null object
dtypes: float64(3), int64(4), object(7)
memory usage: 143.2+ KB


In [3]:
# Identify missing data

missing_data_counts = pd.DataFrame(df.isnull().sum(), columns=['Missing Values'])
missing_data_counts

Missing Values
pclass                  0
survived                0
name                    0
sex                     0
age                   263
sibsp                   0
parch                   0
ticket                  0
fare                    1
cabin                1014
embarked                2
boat                  823
body                 1188
home.dest             564

Observations: 
1. Missing 263 Age records (solve by replacing with average)
2. Missing 1014 Cabin records (doesn't make sense to randomly assign, so remove column )
3. Missing 564 Home port / Destination records (does not impact survival, so remove column)

In [4]:
df['age'].fillna(df['age'].mean(),inplace=True)

In [5]:
df['fare'].fillna(df['fare'].mean(),inplace=True)

In [6]:
df.drop('cabin',axis=1 , inplace=True)

In [7]:
df.drop('home.dest', axis=1, inplace=True)

In [8]:
df.drop('name', axis=1, inplace=True)

In [9]:
df.drop('ticket', axis=1, inplace=True)

In [10]:
df.drop('embarked', axis=1, inplace=True)

In [11]:
# drop duplicates

df.drop_duplicates()

df.head(3)

pclass  survived     sex      age  sibsp  parch      fare boat  body
0       1         1  female  29.0000      0      0  211.3375    2   NaN
1       1         1    male   0.9167      1      2  151.5500   11   NaN
2       1         0  female   2.0000      1      2  151.5500  NaN   NaN

# Categorize 

In [12]:
df.head(3)

pclass  survived     sex      age  sibsp  parch      fare boat  body
0       1         1  female  29.0000      0      0  211.3375    2   NaN
1       1         1    male   0.9167      1      2  151.5500   11   NaN
2       1         0  female   2.0000      1      2  151.5500  NaN   NaN

Notes on Data Types (Categories):

Categorical:
- pclass
- survived
- sex

Numerical: 
- Fare
- Age 

Next steps on Data Preparation: 
- Convert gender to numeric variables 
- Convert age to categories (bins)
- Create new category to describe whether passenger traveled with family


In [13]:
# Convert Gender to Numeric Variable for Analysis

def score_to_numeric(x):
    if x=='female':
        return 1
    if x=='male':
        return 0

In [14]:
# Apply Function to a new Gender Column

df['gender']=df['sex'].apply(score_to_numeric)
df.head(3)

pclass  survived     sex      age  sibsp  parch      fare boat  body  \
0       1         1  female  29.0000      0      0  211.3375    2   NaN   
1       1         1    male   0.9167      1      2  151.5500   11   NaN   
2       1         0  female   2.0000      1      2  151.5500  NaN   NaN   

   gender  
0       1  
1       0  
2       1

In [15]:
# Create a family category

import numpy as np

df['family'] = np.where((df['parch']> 0)|(df['sibsp']> 0),1,0)
df.head(3)



pclass  survived     sex      age  sibsp  parch      fare boat  body  \
0       1         1  female  29.0000      0      0  211.3375    2   NaN   
1       1         1    male   0.9167      1      2  151.5500   11   NaN   
2       1         0  female   2.0000      1      2  151.5500  NaN   NaN   

   gender  family  
0       1       0  
1       0       1  
2       1       1

In [16]:
# create bins for age

bins = [0,18,60,100]

# Group Names 

group_names = ['Child','Adult','Senior']

In [17]:
df['ageCategory']= pd.cut(df['age'], bins, labels=group_names)
df

pclass  survived     sex        age  sibsp  parch      fare boat   body  \
0          1         1  female  29.000000      0      0  211.3375    2    NaN   
1          1         1    male   0.916700      1      2  151.5500   11    NaN   
2          1         0  female   2.000000      1      2  151.5500  NaN    NaN   
3          1         0    male  30.000000      1      2  151.5500  NaN  135.0   
4          1         0  female  25.000000      1      2  151.5500  NaN    NaN   
5          1         1    male  48.000000      0      0   26.5500    3    NaN   
6          1         1  female  63.000000      1      0   77.9583   10    NaN   
7          1         0    male  39.000000      0      0    0.0000  NaN    NaN   
8          1         1  female  53.000000      2      0   51.4792    D    NaN   
9          1         0    male  71.000000      0      0   49.5042  NaN   22.0   
10         1         0    male  47.000000      1      0  227.5250  NaN  124.0   
11         1         1  female  18.000000      1      0  227.5250    4    NaN   
12         1         1  female  24.000000      0      0   69.3000    9    NaN   
13         1         1  female  26.000000      0      0   78.8500    6    NaN   
14         1         1    male  80.000000      0      0   30.0000    B    NaN   
15         1         0    male  29.881135      0      0   25.9250  NaN    NaN   
16         1         0    male  24.000000      0      1  247.5208  NaN    NaN   
17         1         1  female  50.000000      0      1  247.5208    6    NaN   
18         1         1  female  32.000000      0      0   76.2917    8    NaN   
19         1         0    male  36.000000      0      0   75.2417    A    NaN   
20         1         1    male  37.000000      1      1   52.5542    5    NaN   
21         1         1  female  47.000000      1      1   52.5542    5    NaN   
22         1         1    male  26.000000      0      0   30.0000    5    NaN   
23         1         1  female  42.000000      0      0  227.5250    4    NaN   
24         1         1  female  29.000000      0      0  221.7792    8    NaN   
25         1         0    male  25.000000      0      0   26.0000  NaN  148.0   
26         1         1    male  25.000000      1      0   91.0792    7    NaN   
27         1         1  female  19.000000      1      0   91.0792    7    NaN   
28         1         1  female  35.000000      0      0  135.6333    8    NaN   
29         1         1    male  28.000000      0      0   26.5500    D    NaN   
...      ...       ...     ...        ...    ...    ...       ...  ...    ...   
1279       3         0  female  14.000000      0      0    7.8542  NaN    NaN   
1280       3         0    male  22.000000      0      0    7.8958  NaN    NaN   
1281       3         0    male  22.000000      0      0    9.0000  NaN    NaN   
1282       3         0    male  29.881135      0      0    8.0500  NaN    NaN   
1283       3         0    male  29.881135      0      0    7.5500  NaN    NaN   
1284       3         0    male  29.881135      0      0    8.0500  NaN    NaN   
1285       3         0    male  32.500000      0      0    9.5000  NaN  298.0   
1286       3         1  female  38.000000      0      0    7.2292    C    NaN   
1287       3         0    male  51.000000      0      0    7.7500  NaN    NaN   
1288       3         0    male  18.000000      1      0    6.4958  NaN  314.0   
1289       3         0    male  21.000000      1      0    6.4958  NaN    NaN   
1290       3         1  female  47.000000      1      0    7.0000  NaN    NaN   
1291       3         0    male  29.881135      0      0    8.7125  NaN    NaN   
1292       3         0    male  29.881135      0      0    7.5500  NaN    NaN   
1293       3         0    male  29.881135      0      0    8.0500  NaN    NaN   
1294       3         0    male  28.500000      0      0   16.1000  NaN   14.0   
1295       3         0    male  21.000000      0      0    7.2500  NaN    NaN   
1296       3         0    male  27.00

In [18]:
# Categorize survival status as string

import numpy as np

df['survivalStatus'] = np.where((df['survived']> 0),'Survived','Died')
df.head(3)


pclass  survived     sex      age  sibsp  parch      fare boat  body  \
0       1         1  female  29.0000      0      0  211.3375    2   NaN   
1       1         1    male   0.9167      1      2  151.5500   11   NaN   
2       1         0  female   2.0000      1      2  151.5500  NaN   NaN   

   gender  family ageCategory survivalStatus  
0       1       0       Adult       Survived  
1       0       1       Child       Survived  
2       1       1       Child           Died

# Analysis

In [19]:
# pivot table looking at survival by age

pd.pivot_table(df, columns=['survivalStatus','sex'], values=['survived'],index=['ageCategory'], aggfunc='count')

survived                   
survivalStatus     Died      Survived     
sex              female male   female male
ageCategory                               
Adult                97  589      270  127
Child                29   69       63   32
Senior                1   24        6    2

In [20]:
# Pivot table looking at survival by gender

pd.pivot_table(df, columns=['survivalStatus'], values=['survived'], index=['sex'], aggfunc='count')

survived         
survivalStatus     Died Survived
sex                             
female              127      339
male                682      161

In [21]:
# pivot table looking at survival by gender and age

pd.pivot_table(df, columns=['survivalStatus'], values=['survived'],index=['sex','ageCategory'], aggfunc='count')

survived         
survivalStatus         Died Survived
sex    ageCategory                  
female Adult             97      270
       Child             29       63
       Senior             1        6
male   Adult            589      127
       Child             69       32
       Senior            24        2

In [22]:
# pivot table looking at survival by passenger class and gender

pd.pivot_table(df, columns=['survivalStatus'], values=['survived'],index=['sex','pclass'], aggfunc='count')

survived         
survivalStatus     Died Survived
sex    pclass                   
female 1              5      139
       2             12       94
       3            110      106
male   1            118       61
       2            146       25
       3            418       75

Observations: 
- Senior men had a high risk of death
- Female children had a higher rate of survival than male children
- Adult women were more likely to survive than adult men
- Men in third class were less likely to survive than women in third class
- Women in third class had a near 50 / 50 chance of survival, whereas nearly all women in first class and second class survived

In [23]:
# Correlation

df.corr()

pclass  survived       age     sibsp     parch      fare  \
pclass    1.000000 -0.312469 -0.366370  0.060832  0.018322 -0.558477   
survived -0.312469  1.000000 -0.050199 -0.027825  0.082660  0.244208   
age      -0.366370 -0.050199  1.000000 -0.190747 -0.130872  0.171521   
sibsp     0.060832 -0.027825 -0.190747  1.000000  0.373587  0.160224   
parch     0.018322  0.082660 -0.130872  0.373587  1.000000  0.221522   
fare     -0.558477  0.244208  0.171521  0.160224  0.221522  1.000000   
body     -0.034642       NaN  0.059059 -0.099961  0.051099 -0.042665   
gender   -0.124617  0.528693 -0.057398  0.109609  0.213125  0.185484   
family   -0.147393  0.201719 -0.116675  0.591077  0.549022  0.274826   

              body    gender    family  
pclass   -0.034642 -0.124617 -0.147393  
survived       NaN  0.528693  0.201719  
age       0.059059 -0.057398 -0.116675  
sibsp    -0.099961  0.109609  0.591077  
parch     0.051099  0.213125  0.549022  
fare     -0.042665  0.185484  0.274826  
body      1.000000  0.015903  0.002237  
gender    0.015903  1.000000  0.284537  
family    0.002237  0.284537  1.000000

Obervations: 

Strongest correlations are between survival and: 
- Gender (.5286)
- Pclass (.312)
- family (.20)

In [26]:
# Regression with intercept

import statsmodels.formula.api as sm
result = sm.ols(formula='survived ~ gender + pclass + family', data=df).fit()
result.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:               survived   R-squared:                       0.342
Model:                            OLS   Adj. R-squared:                  0.340
Method:                 Least Squares   F-statistic:                     226.0
Date:                Sat, 24 Jun 2017   Prob (F-statistic):          4.71e-118
Time:                        00:37:51   Log-Likelihood:                -638.72
No. Observations:                1309   AIC:                             1285.
Df Residuals:                    1305   BIC:                             1306.
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      0.5242      0.036     14.730      0.000       0.454       0.594
gender         0.4978      0.024     20.861      0.000       0.451       0.545
pclass        -0.1436      0.013    -10.862      0.000      -0.170      -0.118
family         0.0255      0.023      1.087      0.277      -0.020       0.071
==============================================================================
Omnibus:                       59.258   Durbin-Watson:                   1.757
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               66.784
Skew:                           0.553   Prob(JB):                     3.15e-15
Kurtosis:                       2.959   Cond. No.                         9.48
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

Note: P value on family is too high, will rereun without

In [27]:
# Regression with intercept and no family

import statsmodels.formula.api as sm
result = sm.ols(formula='survived ~ gender + pclass', data=df).fit()
result.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:               survived   R-squared:                       0.341
Model:                            OLS   Adj. R-squared:                  0.340
Method:                 Least Squares   F-statistic:                     338.3
Date:                Sat, 24 Jun 2017   Prob (F-statistic):          4.09e-119
Time:                        00:41:19   Log-Likelihood:                -639.31
No. Observations:                1309   AIC:                             1285.
Df Residuals:                    1306   BIC:                             1300.
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      0.5357      0.034     15.761      0.000       0.469       0.602
gender         0.5048      0.023     21.978      0.000       0.460       0.550
pclass        -0.1453      0.013    -11.066      0.000      -0.171      -0.120
==============================================================================
Omnibus:                       60.521   Durbin-Watson:                   1.760
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               68.392
Skew:                           0.560   Prob(JB):                     1.41e-15
Kurtosis:                       2.963   Cond. No.                         9.02
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [28]:
# regression with 

import statsmodels.formula.api as sm
result = sm.ols(formula='survived ~ gender + pclass -1', data=df).fit()
result.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:               survived   R-squared:                       0.515
Model:                            OLS   Adj. R-squared:                  0.515
Method:                 Least Squares   F-statistic:                     695.2
Date:                Sat, 24 Jun 2017   Prob (F-statistic):          2.30e-206
Time:                        00:46:18   Log-Likelihood:                -753.27
No. Observations:                1309   AIC:                             1511.
Df Residuals:                    1307   BIC:                             1521.
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
gender         0.6319      0.023     26.941      0.000       0.586       0.678
pclass         0.0444      0.006      7.745      0.000       0.033       0.056
==============================================================================
Omnibus:                       26.563   Durbin-Watson:                   1.515
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               28.162
Skew:                           0.326   Prob(JB):                     7.67e-07
Kurtosis:                       3.304   Cond. No.                         4.90
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""